### ISSUE TITLE
Snapshot weight averaging (def squash())

### ISSUE THREAD

**Issue by rfratila** [Aifred-Health/Vulcan2#52]
similar to this: https://github.com/fastai/fastai/pull/276/files#diff-6b218dc5c075483a5c9b7a5085e4570dR32

**Comment by rfratila** [comment]
Initial implementation
```
def squash(self, in_place=False):
        """
        Squash together the corresponding weights of all internal snapshots.

        Weighted average of all the model weights in one model. Inspired by
        stochastic weight averaging [1].

        [1] - Averaging Weights Leads to Wider Optima and Better Generalization
         Pavel Izmailov, Dmitrii Podoprikhin, Timur Garipov, Dmitry Vetrov,
         Andrew Gordon Wilson
         https://arxiv.org/abs/1803.05407
         2018

        Parameters
        ----------
        in_place : boolean
            Whether to replace the snapshots with the squashed
            (i.e. averaged) model.

        Returns
        -------
        squashed_network : type(self.template_network)

        """
        # TODO: implement in_place replacement
        if in_place:
            raise NotImplementedError
        temp_net = deepcopy(self.network[0])
        temp_params = temp_net.parameters()
        for net in self.network[1:]:
            snap_params = net.parameters()
            for temp_param, snap_param in zip(temp_params, snap_params):
                temp_param.data *= self.n_snapshots
                temp_param.data += snap_param.data
                temp_param.data /= (self.n_snapshots + 1)
        # TODO: Check if BatchNorm and fix learned params as in
        # https://github.com/timgaripov/swa/blob/master/utils.py#L107
        return temp_net
```

**Comment by nirtiac** [comment]
@rfratila from https://arxiv.org/pdf/1602.05629.pdf, just in case, although you're probably already aware, we need to pay attention to our random initialization. 

> "Recent work indicates that in practice, the loss surfaces of
> sufficiently over-parameterized NNs are surprisingly wellbehaved and in particular less prone to bad local minima than previously thought [11, 17, 9]. And indeed, when we
> start two models from the same random initialization and
> then again train each independently on a different subset of
> the data (as described above), we find that naive parameter
> averaging works surprisingly well (Figure 1, right): the average of these two models, 1
> 2w+
> 1
> 2w
> 0
> , achieves significantly
> lower loss on the full MNIST training set than the best
> model achieved by training on either of the small datasets
> independently. While Figure 1 starts from a random initialization, note a shared starting model wt is used for each
> round of FedAvg, and so the same intuition applies."
